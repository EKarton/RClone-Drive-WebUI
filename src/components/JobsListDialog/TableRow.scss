.table-row {
  &__file-type-cell {
    width: 2rem;
  }

  &__dir-path-cell,
  &__file-name-cell {
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
    padding-right: 1rem;
  }

  &__file-size-cell {
    white-space: nowrap;
    width: 5rem;
  }

  &__status-cell {
    width: 5rem;
  }

  &__cancel-button-cell {
    width: 3rem;
  }
}
