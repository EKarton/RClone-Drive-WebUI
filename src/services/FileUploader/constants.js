export const ActionTypes = Object.freeze({
  UPLOAD_FILE: 'UPLOAD_FILE',
  CANCEL_UPLOAD: 'CANCEL_UPLOAD',
});
