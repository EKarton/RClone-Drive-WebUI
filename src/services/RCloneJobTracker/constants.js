export const ActionTypes = Object.freeze({
  TRACK_NEW_JOB: 'TRACK_NEW_JOB',
  UNTRACK_NEW_JOB: 'UNTRACK_NEW_JOB',
});
