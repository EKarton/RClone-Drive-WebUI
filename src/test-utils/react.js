import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { createMemoryHistory } from 'history';
import { Router } from 'react-router-dom';
import { ColorModeProvider } from 'contexts/ColorMode/index';
import { FileUploadCountsProvider } from 'contexts/FileUploadCounts/index';
import { RCloneInfoProvider } from 'contexts/RCloneInfo';
import { RecentPicturesProvider } from 'contexts/RecentPicturesList';

const customRender = (
  component,
  { initialRCloneInfoState, initialRecentPicturesState } = {},
  { route = '/' } = {}
) => {
  const history = createMemoryHistory({ initialEntries: [route] });

  const renderedComponent = render(
    <RCloneInfoProvider defaultState={initialRCloneInfoState}>
      <RecentPicturesProvider defaultState={initialRecentPicturesState}>
        <Router location={history.location} navigator={history}>
          <ColorModeProvider>
            <FileUploadCountsProvider>{component}</FileUploadCountsProvider>
          </ColorModeProvider>
        </Router>
      </RecentPicturesProvider>
    </RCloneInfoProvider>
  );

  return {
    ...renderedComponent,
    history,
  };
};

export * from '@testing-library/react';
export * from '@testing-library/user-event';
export { customRender, userEvent, render };
