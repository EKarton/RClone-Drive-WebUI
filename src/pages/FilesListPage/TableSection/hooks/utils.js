import { JobStatus, JobTypes } from 'utils/constants';

export function getFileNames(data) {
  return new Set(data.filter((file) => !file.IsDir).map((file) => file.Name));
}

export function getFolderNames(data) {
  return new Set(data.filter((file) => file.IsDir).map((file) => file.Name));
}

export function getUploadingJobs(jobs, remote, path) {
  return jobs.filter(
    (job) =>
      job.jobType === JobTypes.UPLOAD_FILE &&
      job.remote === remote &&
      job.dirPath.startsWith(path) &&
      job.status.value !== JobStatus.SUCCESS
  );
}

export function getUploadingFiles(jobs, remote, path) {
  return getUploadingJobs(jobs, remote, path).filter((job) => job.dirPath === path);
}

export function getUploadingFolders(jobs, remote, path) {
  return getUploadingJobs(jobs, remote, path).filter((job) => job.dirPath !== path);
}

export function getMovingJobs(jobs, remote, path) {
  const supportedJobTypes = new Set([
    JobTypes.MOVE_FILE,
    JobTypes.MOVE_FOLDER,
    JobTypes.RENAME_FILE,
    JobTypes.RENAME_FOLDER,
  ]);

  return jobs.filter((job) => {
    const correctJob = supportedJobTypes.has(job.jobType);
    const inProgress = job.status.value === JobStatus.ONGOING;

    if (!correctJob || !inProgress) {
      return false;
    }

    const isMatchSrc = job.src?.remote === remote && job.src?.dirPath === path;
    const isMatchTarget = job.target?.remote === remote && job.target?.dirPath === path;

    return correctJob && inProgress && (isMatchSrc || isMatchTarget);
  });
}
