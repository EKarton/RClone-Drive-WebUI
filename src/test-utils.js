import { FileViewerProvider, initialState } from 'contexts/FileViewerStore';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { createMemoryHistory } from 'history';
import { Router } from 'react-router';

const customRender = (
  component,
  { initialFileViewerState = initialState } = {},
  { route = '/' } = {}
) => {
  const history = createMemoryHistory({ initialEntries: [route] });

  return render(
    <FileViewerProvider defaultState={initialFileViewerState}>
      <Router history={history}>{component}</Router>
    </FileViewerProvider>
  );
};

export * from '@testing-library/react';
export * from '@testing-library/user-event';
export { customRender, userEvent };
