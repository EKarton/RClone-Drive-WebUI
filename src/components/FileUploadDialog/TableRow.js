import CancelIcon from '@mui/icons-material/Cancel';
import IconButton from '@mui/material/IconButton';
import TableCell from '@mui/material/TableCell';
import MuiTableRow from '@mui/material/TableRow';
import prettyBytes from 'pretty-bytes';
import { useEffect, useState } from 'react';
import { UploadStatusTypes } from 'utils/constants';
import './TableRow.scss';
import TableRowIcon from './TableRowIcon';

export default function TableRow({ file }) {
  const [status, setStatus] = useState(file.status.value);

  useEffect(() => {
    const subscriber = file.status.subscribe((newStatus) => {
      setStatus(newStatus);
    });

    return () => {
      subscriber.unsubscribe();
    };
  }, [file]);

  const handleCancelButtonClick = () => {
    file.cancelUpload();
  };

  const dirPathWithRemote = `${file.remote}:${file.dirPath}`;

  return (
    <MuiTableRow sx={{ '&:last-child td, &:last-child th': { border: 0 } }}>
      <TableCell className="table-row__file-type-cell">
        <TableRowIcon fileType={file.type} />
      </TableCell>
      <TableCell classes={{ root: 'table-row__file-name-cell' }}>{file.name}</TableCell>
      <TableCell classes={{ root: 'table-row__dir-path-cell' }}>
        {dirPathWithRemote}
      </TableCell>
      <TableCell align="right" className="table-row__file-size-cell">
        {prettyBytes(file.size)}
      </TableCell>
      <TableCell align="right" className="table-row__status-cell">
        {status}
      </TableCell>
      <TableCell align="right" className="table-row__cancel-button-cell">
        <IconButton
          aria-label="cancel upload"
          onClick={handleCancelButtonClick}
          disabled={status !== UploadStatusTypes.UPLOADING}
          data-testid="cancel-upload-button"
        >
          <CancelIcon />
        </IconButton>
      </TableCell>
    </MuiTableRow>
  );
}
