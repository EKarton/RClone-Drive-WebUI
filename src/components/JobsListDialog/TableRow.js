import CancelIcon from '@mui/icons-material/Cancel';
import IconButton from '@mui/material/IconButton';
import TableCell from '@mui/material/TableCell';
import MuiTableRow from '@mui/material/TableRow';
import PropTypes from 'prop-types';
import { useEffect, useState } from 'react';
import { BehaviorSubject } from 'rxjs';
import { JobStatus } from 'utils/constants';
import './TableRow.scss';
import TableRowIcon from './TableRowIcon';
import TableRowDescription from './TableRowJobDescription';

export default function TableRow({ job, ...props }) {
  const [status, setStatus] = useState(null);

  useEffect(() => {
    const subscriber = job.status.subscribe((newStatus) => {
      setStatus(newStatus);
    });

    return () => {
      subscriber.unsubscribe();
    };
  }, [job]);

  const handleCancelButtonClick = () => {
    job.cancelJob();
  };

  return (
    <MuiTableRow sx={{ '&:last-child td, &:last-child th': { border: 0 } }} {...props}>
      <TableCell className="table-row__file-type-cell">
        <TableRowIcon jobType={job.jobType} />
      </TableCell>
      <TableCell classes={{ root: 'table-row__file-name-cell' }}>
        <TableRowDescription status={status} job={job} />
      </TableCell>
      <TableCell align="right" className="table-row__status-cell">
        {status}
      </TableCell>
      <TableCell align="right" className="table-row__cancel-button-cell">
        <IconButton
          aria-label="cancel job"
          onClick={handleCancelButtonClick}
          disabled={status !== JobStatus.ONGOING}
          data-testid="cancel-job-button"
        >
          <CancelIcon />
        </IconButton>
      </TableCell>
    </MuiTableRow>
  );
}

TableRow.propTypes = {
  file: PropTypes.shape({
    remote: PropTypes.string,
    dirPath: PropTypes.string,
    name: PropTypes.string,
    type: PropTypes.string,
    status: PropTypes.instanceOf(BehaviorSubject),
  }),
};
