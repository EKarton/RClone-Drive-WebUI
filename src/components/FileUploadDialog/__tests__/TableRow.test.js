import { BehaviorSubject } from 'rxjs';
import { UploadStatusTypes } from 'utils/constants';
import { render, act, userEvent, screen } from 'test-utils/react';
import TableRow from '../TableRow';

describe('TableRow', () => {
  it('should match snapshot given file is being uploaded', () => {
    const file = createBaseFile();
    const { baseElement } = render(<TableRow file={file} />);

    expect(baseElement).toMatchSnapshot();
  });

  it('should update snapshot when file status has changed given file is being uploaded', () => {
    const file = createBaseFile();
    const { baseElement } = render(<TableRow file={file} />);

    act(() => file.status.next(UploadStatusTypes.SUCCESS));

    expect(baseElement).toMatchSnapshot();
  });

  it('should call file.cancelUpload() when user clicks on cancel button given file is being uploaded', async () => {
    const file = createBaseFile();
    render(<TableRow file={file} />);

    userEvent.click(screen.getByTestId('cancel-upload-button'));

    expect(file.cancelUpload).toBeCalled();
  });

  const createBaseFile = () => ({
    remote: 'gdrive',
    dirPath: 'Documents',
    name: 'Dog.png',
    size: 10000,
    type: 'application/jpg',
    status: new BehaviorSubject(UploadStatusTypes.UPLOADING),
    cancelUpload: jest.fn(),
  });
});
