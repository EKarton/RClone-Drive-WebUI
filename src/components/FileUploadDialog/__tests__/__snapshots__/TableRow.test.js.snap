// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableRow should match snapshot given file is being uploaded 1`] = `
<body>
  <div>
    <tr
      class="MuiTableRow-root css-34nofg-MuiTableRow-root"
    >
      <td
        class="MuiTableCell-root MuiTableCell-sizeMedium table-row__file-type-cell css-bsph2u-MuiTableCell-root"
      >
        <span
          aria-label="application/jpg"
          class=""
          data-mui-internal-clone-element="true"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="DescriptionIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M14 2H6c-1.1 0-1.99.9-1.99 2L4 20c0 1.1.89 2 1.99 2H18c1.1 0 2-.9 2-2V8l-6-6zm2 16H8v-2h8v2zm0-4H8v-2h8v2zm-3-5V3.5L18.5 9H13z"
            />
          </svg>
        </span>
      </td>
      <td
        class="table-row__file-name-cell MuiTableCell-root MuiTableCell-sizeMedium css-bsph2u-MuiTableCell-root"
      >
        Dog.png
      </td>
      <td
        class="table-row__dir-path-cell MuiTableCell-root MuiTableCell-sizeMedium css-bsph2u-MuiTableCell-root"
      >
        gdrive:Documents
      </td>
      <td
        class="MuiTableCell-root MuiTableCell-alignRight MuiTableCell-sizeMedium table-row__file-size-cell css-r7oxtk-MuiTableCell-root"
      >
        10 kB
      </td>
      <td
        class="MuiTableCell-root MuiTableCell-alignRight MuiTableCell-sizeMedium table-row__status-cell css-r7oxtk-MuiTableCell-root"
      >
        Uploading
      </td>
      <td
        class="MuiTableCell-root MuiTableCell-alignRight MuiTableCell-sizeMedium table-row__cancel-button-cell css-r7oxtk-MuiTableCell-root"
      >
        <button
          aria-label="cancel upload"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
          data-testid="cancel-upload-button"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="CancelIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2zm5 13.59L15.59 17 12 13.41 8.41 17 7 15.59 10.59 12 7 8.41 8.41 7 12 10.59 15.59 7 17 8.41 13.41 12 17 15.59z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </td>
    </tr>
  </div>
</body>
`;

exports[`TableRow should update snapshot when file status has changed given file is being uploaded 1`] = `
<body>
  <div>
    <tr
      class="MuiTableRow-root css-34nofg-MuiTableRow-root"
    >
      <td
        class="MuiTableCell-root MuiTableCell-sizeMedium table-row__file-type-cell css-bsph2u-MuiTableCell-root"
      >
        <span
          aria-label="application/jpg"
          class=""
          data-mui-internal-clone-element="true"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="DescriptionIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M14 2H6c-1.1 0-1.99.9-1.99 2L4 20c0 1.1.89 2 1.99 2H18c1.1 0 2-.9 2-2V8l-6-6zm2 16H8v-2h8v2zm0-4H8v-2h8v2zm-3-5V3.5L18.5 9H13z"
            />
          </svg>
        </span>
      </td>
      <td
        class="table-row__file-name-cell MuiTableCell-root MuiTableCell-sizeMedium css-bsph2u-MuiTableCell-root"
      >
        Dog.png
      </td>
      <td
        class="table-row__dir-path-cell MuiTableCell-root MuiTableCell-sizeMedium css-bsph2u-MuiTableCell-root"
      >
        gdrive:Documents
      </td>
      <td
        class="MuiTableCell-root MuiTableCell-alignRight MuiTableCell-sizeMedium table-row__file-size-cell css-r7oxtk-MuiTableCell-root"
      >
        10 kB
      </td>
      <td
        class="MuiTableCell-root MuiTableCell-alignRight MuiTableCell-sizeMedium table-row__status-cell css-r7oxtk-MuiTableCell-root"
      >
        Success
      </td>
      <td
        class="MuiTableCell-root MuiTableCell-alignRight MuiTableCell-sizeMedium table-row__cancel-button-cell css-r7oxtk-MuiTableCell-root"
      >
        <button
          aria-label="cancel upload"
          class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
          data-testid="cancel-upload-button"
          disabled=""
          tabindex="-1"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="CancelIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2zm5 13.59L15.59 17 12 13.41 8.41 17 7 15.59 10.59 12 7 8.41 8.41 7 12 10.59 15.59 7 17 8.41 13.41 12 17 15.59z"
            />
          </svg>
        </button>
      </td>
    </tr>
  </div>
</body>
`;
